ws/connection/connection.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2020 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package connection
+
+import (
+	"net/http"
+	"net/url"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/vmihailenco/msgpack"
+
+	"github.com/mendersoftware/go-lib-micro/ws"
+)
+
+type Connection struct {
+	writeMutex sync.Mutex
+	// the connection handler
+	connection *websocket.Conn
+	// Time allowed to write a message to the peer.
+	writeWait time.Duration
+	// Maximum message size allowed from peer.
+	maxMessageSize int
+	// Time allowed to read the next pong message from the peer.
+	defaultPingWait time.Duration
+}
+
+//Websocket connection routine. setup the ping-pong and connection settings
+func NewConnection(u url.URL,
+	token string,
+	writeWait time.Duration,
+	maxMessageSize int64,
+	defaultPingWait time.Duration) (*Connection, error) {
+	var ws *websocket.Conn
+	dialer := *websocket.DefaultDialer
+
+	headers := http.Header{}
+	headers.Set("Authorization", "Bearer "+token)
+	ws, _, err := dialer.Dial(u.String(), headers)
+	if err != nil {
+		return nil, err
+	}
+
+	// ping-pong
+	ws.SetReadLimit(maxMessageSize)
+	ws.SetReadDeadline(time.Now().Add(defaultPingWait))
+	ws.SetPingHandler(func(message string) error {
+		pongWait, _ := strconv.Atoi(message)
+		ws.SetReadDeadline(time.Now().Add(time.Duration(pongWait) * time.Second))
+		return ws.WriteControl(websocket.PongMessage, []byte{}, time.Now().Add(writeWait))
+	})
+	return &Connection{
+		connection:      ws,
+		writeWait:       0,
+		maxMessageSize:  0,
+		defaultPingWait: 0,
+	}, nil
+}
+
+func (c *Connection) WriteMessage(m *ws.ProtoMsg) (err error) {
+	data, err := msgpack.Marshal(m)
+	if err != nil {
+		return err
+	}
+	c.writeMutex.Lock()
+	err = c.connection.SetWriteDeadline(time.Now().Add(c.writeWait))
+	err = c.connection.WriteMessage(websocket.BinaryMessage, data)
+	c.writeMutex.Unlock()
+	return nil
```

Comment:
Always defer mutex unlocks in case the goroutine panics. `err` variable assigned but not used.
```suggestion
	c.writeMutex.Lock()
	defer c.writeMutex.Unlock()
	err = c.connection.SetWriteDeadline(time.Now().Add(c.writeWait))
	if err != nil {
		return err
	}
	err = c.connection.WriteMessage(websocket.BinaryMessage, data)
	return err
```

Reply:
if you choose to return in the critical section, yes, but in that case defer is redundant, but I will add it. and if it panics in here it is os.Exit basically. perhaps we can ignore the error on SetWriteDeadLine.